lib/Onnxifi/GlowOnnxifiManager.h

```diff
@@ -0,0 +1,115 @@
+  /// ever be one GlowOnnxifiManager.
+  static GlowOnnxifiManager &get();
+
+  /// Disallow copying and moving GlowOnnxifiManager to help enforce singleton
+  /// pattern.
+  GlowOnnxifiManager(const GlowOnnxifiManager &) = delete;
+  GlowOnnxifiManager(GlowOnnxifiManager &&) = delete;
+  GlowOnnxifiManager &operator=(const GlowOnnxifiManager &) = delete;
+  GlowOnnxifiManager &operator=(GlowOnnxifiManager &&) = delete;
+
+  /// Add a new glow BackendId \p backendId to the set of valid BackendIds.
+  /// GlowOnnxifiManager then owns this BackendId and is responsible for
+  /// deallocating when it is released. Can be called safely by multiple threads
+  /// concurrently.
+  void addBackendId(BackendIdPtr backendId);
+
+  /// Create a new glow Backend associated with \p backendId.
+  /// Can be called safely by multiple threads concurrently.
+  BackendPtr createBackend(BackendIdPtr backendId);
+
+  /// Create a new glow Event.
+  /// Can be called safely by multiple threads concurrently.
+  EventPtr createEvent();
+
+  /// Create a new glow Graph associated with \p backend.
+  /// Can be called safely by multiple threads concurrently.
+  GraphPtr createGraph(BackendPtr backend);
+
+  /// Check if \p backendId is a BackendId created and managed by glow.
+  /// Can be called safely by multiple threads concurrently.
+  bool isValid(BackendIdPtr backendId);
+
+  /// Check if \p backend is a Backend created and managed by glow.
+  /// Can be called safely by multiple threads concurrently.
+  bool isValid(BackendPtr backend);
+
+  /// Check if \p event is a Event created and managed by glow.
+  /// Can be called safely by multiple threads concurrently.
+  bool isValid(EventPtr event);
+
+  /// Check if \p graph is a Graph created and managed by glow.
+  /// Can be called safely by multiple threads concurrently.
+  bool isValid(GraphPtr graph);
+
+  /// Free \p backendId.
+  /// Can be called safely by multiple threads concurrently.
+  void release(BackendIdPtr backendId);
+
+  /// Free \p backend.
+  /// Can be called safely by multiple threads concurrently.
+  void release(BackendPtr backend);
+
+  /// Free \p event.
+  /// Can be called safely by multiple threads concurrently.
+  void release(EventPtr event);
+
+  /// Free \p graph.
+  /// Can be called safely by multiple threads concurrently.
+  void release(GraphPtr graph);
+
+private:
+  GlowOnnxifiManager() = default;
+
+  /// The set of all valid glow BackendIds.
+  std::unordered_set<BackendIdPtr> backendIds_;
+
+  /// The set of all valid glow Backends.
+  std::unordered_set<BackendPtr> backends_;
+
+  /// The set of all valid glow Events.
+  std::unordered_set<EventPtr> events_;
+
+  /// The set of all valid glow Graphs.
+  std::unordered_set<GraphPtr> graphs_;
+
+  /// Mutex that protects all members of GlowOnnxifiManager.
+  /// TODO: can use one mutex per set if performance becomes an issue.
+  mutable std::mutex m_;
```

Comment:
why do you need this to be mutable?

Reply:
`mutable` for mutexes is usually good (I think) so that you can take a const ref to an object and still call methods that need to lock the mutex

Reply:
we do not have any const methods in this class :)

Reply:
good point, valid() should be const, changed it